storage/s3.go

```diff
@@ -382,6 +386,175 @@ func (s *S3) Get(
+	}()
+
+	decoder := json.NewDecoder(reader)
+	// decoder.DontCopyRawMessage()
+	for {
+		var record json.RawMessage
+		err := decoder.Decode(&record)
+		if err != nil {
+			if err != io.EOF {
+				resultch <- &SelectRecord{
+					Err: err,
+				}
+			}
+			return
+		}
+		resultch <- &SelectRecord{
+			Record: record,
+		}
+	}
+	err := resp.EventStream.Err()
+	if err != nil {
+		resultch <- &SelectRecord{
+			Err: err,
+		}
+	}
+}
+
+func (s *S3) selectWorker(ctx context.Context, input s3.SelectObjectContentInput, objch <-chan *Object, resultch chan<- *SelectRecord) {
+	for object := range objch {
+		// TODO: reintroduce
+		if object.Type.IsDir() { // || errorpkg.IsCancelation(object.Err) {
+			continue
+		}
+
+		if err := object.Err; err != nil {
+			resultch <- &SelectRecord{
+				Err: object.Err,
+			}
+			continue
+		}
+
+		if object.StorageClass.IsGlacier() {
+			// TODO: this isn't fatal - how to signal?
+			resultch <- &SelectRecord{
+				Err: fmt.Errorf("object '%v' is on Glacier storage", object),
+			}
+			continue
+		}
+
+		url := object.URL
+		input.Bucket = aws.String(url.Bucket)
+		input.Key = aws.String(url.Path)
+
+		out, err := s.api.SelectObjectContentWithContext(ctx, &input)
+		if err != nil {
+			resultch <- &SelectRecord{
+				Err: err,
+			}
+		} else {
+			s.selectStream(out, resultch)
+		}
+	}
+}
+
+type Query struct {
+	ExpressionType  string
+	Expression      string
+	CompressionType string
+}
+
+func writev(to io.Writer, vector ...[]byte) (err error) {
+	for _, vec := range vector {
+		_, err = to.Write(vec)
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+func (s *S3) Select(ctx context.Context, objch <-chan *Object, to io.Writer, query *Query, concurrency int) error {
+	if s.dryRun {
+		return nil
+	}
+
+	baseInput := s3.SelectObjectContentInput{
+		ExpressionType: aws.String(query.ExpressionType),
+		Expression:     aws.String(query.Expression),
+		InputSerialization: &s3.InputSerialization{
+			CompressionType: aws.String(query.CompressionType),
+			JSON: &s3.JSONInput{
+				Type: aws.String("Lines"),
+			},
+		},
+		OutputSerialization: &s3.OutputSerialization{
+			JSON: &s3.JSONOutput{},
+		},
+		// RequestProgress: &s3.RequestProgress{
+		// 	// TODO: conditionally enable? for larger objects? ?
+		// 	Enabled: false,
+		// },
+	}
+
+	workerCtx, cancel := context.WithCancel(ctx)
+
+	resultch := make(chan *SelectRecord, concurrency)
+	var wg sync.WaitGroup
+	wg.Add(concurrency)
+	for worker := 0; worker < concurrency; worker++ {
+		go func() {
+			defer wg.Done()
+			s.selectWorker(workerCtx, baseInput, objch, resultch)
+		}()
+	}
+
+	go func() {
+		defer close(resultch)
+		wg.Wait()
+	}()
+	// go func() {
+	// 	s.selectWorker(workerCtx, baseInput, objch, resultch)
+	// 	close(resultch)
+	// }()
+
+	var merror error
+	var fatalError error
+	for record := range resultch {
+		if record.Err == nil {
+			if fatalError != nil {
+				continue
+			}
+			if err := writev(to, record.Record, []byte{'\n'}); err != nil {
```

Comment:
`writev` function looks unnecessary because you can write:

```
_, err := to.Write(append(record.Record, '\n'))
```

Reply:
I guess I'm not terribly sure about this, but my impression was that this could be faster than copying all of `record.Record`. Probably just me (needlessly) microoptimizing.

Reply:
We need to see the generated code to see if this code needs optimization. I don't think it's worth it at this stage.

Reply:
LOL so removing `writev` results in a ~17% speedup for IO-heavy `select` calls on my machine (N=8)